### PR TITLE
Added calculation of the project progress on projects list initializa…

### PR DESCRIPTION
**Pull request by mXaln**
…tion

### CODE REVIEW

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityProjectManager.java [hunk 1]
```diff
@@ -199,6 +199,7 @@ private void initializeViews() {
 
         hideProjectsIfEmpty(mNumProjects);
         if (mNumProjects > 0) {
+            calculateProjectsProgress();
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityProjectManager.java]: I don't want to put this here; there's already too much work going on to start up this activity that we need to make sure that things that won't change very often don't happen all the time.

progress is only "wrong" when the filesystem and database were out of sync, correct?
Then let's add a resultcode to the resync to return whether or not anything changed, and only on change do we calculate all the project progress.

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/database/ProjectDatabaseHelper.java [hunk 2]
```diff
@@ -1140,6 +1140,26 @@ public int getProjectProgressSum(int projectId) {
         return progress;
     }
 
+    public int getProjectProgress(int projectId) {
+        String projectIdString = String.valueOf(projectId);
+        SQLiteDatabase db = getReadableDatabase();
+        final String query = String.format("SELECT %s FROM %s WHERE %s=?",
+                ProjectContract.ProjectEntry.PROJECT_PROGRESS, ProjectContract.ProjectEntry.TABLE_PROJECT, ProjectContract.ProjectEntry._ID);
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/database/ProjectDatabaseHelper.java]: when you have to split arguments onto another line because of length (and this still has line length issues), each parameter goes on its own line like so:

```
 final String query = String.format(
    "SELECT %s FROM %s WHERE %s=?",
    ProjectContract.ProjectEntry.PROJECT_PROGRESS, 
    ProjectContract.ProjectEntry.TABLE_PROJECT, 
    ProjectContract.ProjectEntry._ID
);
```

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/database/ProjectDatabaseHelper.java [hunk 3]
```diff
@@ -1140,6 +1140,26 @@ public int getProjectProgressSum(int projectId) {
...
+        return Math.round(progress);
+    }
+
+    public void setProjectProgress(int projectId, int progress) {
+        final String whereClause = String.format("%s=?", ProjectContract.ProjectEntry._ID);
+        String projectIdString = String.valueOf(projectId);
+        SQLiteDatabase db = getReadableDatabase();
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(ProjectContract.ProjectEntry.PROJECT_PROGRESS, progress);
+        db.update(ProjectContract.ProjectEntry.TABLE_PROJECT, contentValues, whereClause, new String[]{projectIdString});
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/database/ProjectDatabaseHelper.java:1156]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java [hunk 4]
```diff
@@ -192,13 +195,20 @@ private void prepareUnitCardData() {
             List<Chunk> chunks = chunkPlugin.getChapter(mChapterNum).getChunks();
             for (Chunk unit : chunks) {
                 String title = Utils.capitalizeFirstLetter(mProject.getModeName()) + " " + unit.getLabel();
-                mUnitCardList.add(new UnitCard(mAdapter, mProject, title, mChapterNum, unit.getStartVerse(), unit.getEndVerse()));
+                mUnitCardList.add(new UnitCard(mAdapter, mProject, title, mChapterNum, unit.getStartVerse(), unit.getEndVerse(), this));
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java]: line length

Comment (mXaln) [reply translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java]: @sarabiaj Updated

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    public ProjectProgress(Project project, Context context) {
+        this.project = project;
+        this.context = context;
+        this.db = new ProjectDatabaseHelper(context);
+    }
+
+    public int calculateProjectProgress() {
+        try {
+            ChunkPlugin chunks = project.getChunkPlugin(new ChunkPluginLoader(context));
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java]: why not do this in the constructor, and move it to the top as a member variable since most every method here uses it?

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            List<Chapter> projectChapters = chunks.getChapters();
+            for (Chapter chapter: projectChapters) {
+                updateChapterProgress(chapter);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public Chapter getChapter(int chapterNumber) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java]: this really shouldn't be a public API for a progress use case object

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java (hunk 5 above)
Comment (mXaln) [reply translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java]: @sarabiaj That's a good idea. Updated!

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java (hunk 6 above)
Comment (mXaln) [reply translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java]: @sarabiaj Updated!

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityProjectManager.java [hunk 7]
```diff
@@ -44,11 +46,9 @@
 import org.wycliffeassociates.translationrecorder.utilities.Task;
 import org.wycliffeassociates.translationrecorder.utilities.TaskFragment;
 
-import java.io.BufferedOutputStream;
-import java.io.File;
-import java.io.FileNotFoundException;
-import java.io.OutputStream;
+import java.io.*;
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityProjectManager.java]: don't import .*

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/widgets/UnitCard.java [hunk 8]
```diff
@@ -74,14 +78,16 @@
     private int mTakeCount;
     private SoftReference<List<File>> mTakeList;
     private SoftReference<AudioPlayer> mAudioPlayer;
+    private OnTakeDeleteListener onTakeDeleteListener;
 
     // Constructors
-    public UnitCard(DatabaseAccessor db, Project project, String title, int chapter, int firstVerse, int endVerse) {
+    public UnitCard(DatabaseAccessor db, Project project, String title, int chapter, int firstVerse, int endVerse, OnTakeDeleteListener listener) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/widgets/UnitCard.java]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        return null;
+    }
+
+    private int calculateProgress(int current, int total) {
+        return Math.round(((float) current / total) * 100);
+    }
+
+    public void destroy() {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java]: Manual resource cleanup is difficult and should probably be considered an antipattern in a garbage collected language especially. 

nulling out these objects isn't necessary either, they will get garbage collected when ProjectProgress gets garbage collected when the Activity is flagged for GC.

Fortunately, because of what we discussed about injecting the database, you wouldn't need to worry about closing (or opening) the database in this class.

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityProjectManager.java (hunk 7 above)
Comment (mXaln) [reply translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityProjectManager.java]: @sarabiaj yeah, I know... It's just my ide does that, and I didn't pay attention

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java (hunk 9 above)
Comment (mXaln) [reply translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectProgress.java]: @sarabiaj Yep, I've just removed this method

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java [hunk 10]
```diff
@@ -34,15 +40,16 @@
 
     Context mCtx;
     FragmentManager mFragmentManager;
+    ProjectDatabaseHelper mDb;
 
     public ProjectListResyncTask(int taskId, Context ctx, FragmentManager fm) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java]: Lets Inject the database here too

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java [hunk 11]
```diff
@@ -187,18 +189,29 @@ public void onNegativeClick(RatingDialog dialog) {
...
+                ));
             }
-        } catch (IOException e) {
-            e.printStackTrace();
         }
     }
 
+    @Override
+    public void onTakeDeleted() {
+        ProjectProgress pp = new ProjectProgress(mProject, db, chunkPlugin.getChapters());
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java]: this can be moved to the top; I'd name it projectProgress over pp though since its use will be further from the declaration. (you're in the unit list, so everything is in the same project, this will cut down on a lot of work being done repeatedly)

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java [hunk 12]
```diff
@@ -34,15 +38,16 @@
 
     Context mCtx;
     FragmentManager mFragmentManager;
+    ProjectDatabaseHelper mDb;
 
-    public ProjectListResyncTask(int taskId, Context ctx, FragmentManager fm) {
+    public ProjectListResyncTask(int taskId, Context ctx, FragmentManager fm, ProjectDatabaseHelper db) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java]: if you inject the chunkpluginloader, you won't need the context object dependency here

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java [hunk 13]
```diff
@@ -130,16 +135,15 @@ public ProjectListResyncTask(int taskId, Context ctx, FragmentManager fm) {
...
     public void run() {
-        ProjectDatabaseHelper db = new ProjectDatabaseHelper(mCtx);
         Map<Project, File> directoriesOnFs = getProjectDirectoriesOnFileSystem();
         //if the number of projects doesn't match up between the filesystem and the db, OR,
         //the projects themselves don't match an id in the db, then resync everything (only resyncing
         // projects missing won't remove dangling take references in the db)
         //NOTE: removing a project only removes dangling takes, not the project itself from the db
-        if (directoriesOnFs.size() != db.getNumProjects() || db.projectsNeedingResync(directoriesOnFs.keySet()).size() > 0) {
-            fullResync(db, directoriesOnFs);
+        if (directoriesOnFs.size() != mDb.getNumProjects() || mDb.projectsNeedingResync(directoriesOnFs.keySet()).size() > 0) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/Recording/RecordingActivity.java [hunk 14]
```diff
@@ -426,6 +431,11 @@ private void addTakeToDb() {
                 0,
                 mUser.getId()
         );
+
+        if(chunkPlugin != null) {
+            ProjectProgress pp = new ProjectProgress(mProject, db, chunkPlugin.getChapters());
+            pp.updateProjectProgress();
+        }
         db.close();
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/Recording/RecordingActivity.java]: probably not good to close this if we're going to have one instance for the whole activity

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/widgets/UnitCard.java [hunk 15]
```diff
@@ -145,7 +162,11 @@ private void refreshAudioPlayer() {
                 ap.loadFile(getTakeList().get(mTakeIndex));
             }
         }
-        ap.refreshView(mViewHolder.elapsed, mViewHolder.duration, mViewHolder.takePlayPauseBtn, mViewHolder.seekBar);
+        ap.refreshView(mViewHolder.elapsed,
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/widgets/UnitCard.java]: first parameter goes on a new line like you did on line 135

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java [hunk 16]
```diff
@@ -187,15 +196,27 @@ public void onNegativeClick(RatingDialog dialog) {
...
 
     private void prepareUnitCardData() {
-        try {
-            ChunkPlugin chunkPlugin = mProject.getChunkPlugin(new ChunkPluginLoader(this));
+        if(chunkPlugin != null) {
             List<Chunk> chunks = chunkPlugin.getChapter(mChapterNum).getChunks();
             for (Chunk unit : chunks) {
                 String title = Utils.capitalizeFirstLetter(mProject.getModeName()) + " " + unit.getLabel();
-                mUnitCardList.add(new UnitCard(mAdapter, mProject, title, mChapterNum, unit.getStartVerse(), unit.getEndVerse()));
+                mUnitCardList.add(new UnitCard(
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/activities/ActivityUnitList.java]: code formatting

File: translationRecorder/app/src/androidTest/java/org/wycliffeassociates/translationrecorder/recordingapp/UnitTests/project/ProjectFileUtilsTest.java [hunk 17]
```diff
@@ -29,10 +29,11 @@
     Book book = new Book("gen", "Genesis", "ot", 1);
     Version version = new Version("ulb", "Unlocked Literal Bible");
     String mask = "1111111111";
+    int sort = 1;
 
     String jarName = "biblechunk.jar";
     String className = "org.wycliffeassociates.translationrecorder.biblechunk.BibleChunkPlugin";
-    Anthology anthology = new Anthology("ot", "Old Testament", "bible", regex, groups, mask, jarName, className);
+    Anthology anthology = new Anthology("ot", "Old Testament", "bible", sort, regex, groups, mask, jarName, className);
```
Comment (jsarabia) [inline translationRecorder/app/src/androidTest/java/org/wycliffeassociates/translationrecorder/recordingapp/UnitTests/project/ProjectFileUtilsTest.java]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/FilesPage/Export/TranslationExchangeExport.java [hunk 18]
```diff
@@ -54,7 +54,8 @@ protected void handleUserInput() {
             @Override
             public void run() {
                 Context ctx = mCtx.getActivity().getApplicationContext();
-                uploadBinary(ctx, outputFile());
+                ProjectDatabaseHelper db = ((TranslationRecorderApp)mCtx.getActivity().getApplication()).getDatabase();
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/FilesPage/Export/TranslationExchangeExport.java]: don't do this; provide it as a dependency

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/MainMenu.java [hunk 19]
```diff
@@ -287,6 +284,8 @@ private void initApp() {
             brd.show(fm, "Bug Report Dialog");
         }
 
+        db = ((TranslationRecorderApp)getApplication()).getDatabase();
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/MainMenu.java]: this is not related to initializing views; maybe move it to onCreate

File: translationRecorder/app/src/main/AndroidManifest.xml [hunk 20]
```diff
@@ -96,6 +96,15 @@
         <service android:name="org.wycliffeassociates.translationrecorder.Recording.WavRecorder" />
         <service android:name="org.wycliffeassociates.translationrecorder.Recording.WavFileWriter" />
 
+        <provider
```
Comment (jsarabia) [inline translationRecorder/app/src/main/AndroidManifest.xml:99]: what is this needed for?

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/ChapterCardAdapter.java [hunk 21]
```diff
@@ -60,10 +61,15 @@
     private int DROPPED_CARD_EMPTY_TEXT_COLOR;
 
     // Constructor
-    public ChapterCardAdapter(AppCompatActivity context, Project project, List<ChapterCard> chapterCardList) {
+    public ChapterCardAdapter(
+            AppCompatActivity context,
+            Project project,
+            List<ChapterCard> chapterCardList,
+            ProjectDatabaseHelper db) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/ChapterCardAdapter.java]: put the ) { on a new line

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/ProjectAdapter.java [hunk 22]
```diff
@@ -71,12 +68,12 @@ public View getView(final int position, View convertView, final ViewGroup parent
             holder = (ViewHolder) convertView.getTag();
         }
 
-        initializeProjectCard(mCtx, mProjectList.get(position), mDb, holder.mLanguage, holder.mBook, holder.mInfo, holder.mRecord, holder.mTextLayout, holder.mProgressPie);
+        initializeProjectCard(mCtx, mProjectList.get(position), db, holder.mLanguage, holder.mBook, holder.mInfo, holder.mRecord, holder.mTextLayout, holder.mProgressPie);
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/ProjectAdapter.java]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/ProjectAdapter.java [hunk 23]
```diff
@@ -71,12 +68,12 @@ public View getView(final int position, View convertView, final ViewGroup parent
...
 
         return convertView;
     }
 
-    public static void initializeProjectCard(final Activity ctx, final Project project, ProjectDatabaseHelper dB, TextView languageView, TextView bookView,
+    public static void initializeProjectCard(final Activity ctx, final Project project, final ProjectDatabaseHelper dB, TextView languageView, TextView bookView,
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/ProjectAdapter.java]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/UnitCardAdapter.java [hunk 24]
```diff
@@ -44,70 +44,64 @@
...
 
 
     // Constructor
-    public UnitCardAdapter(AppCompatActivity context, Project project, int chapter, List<UnitCard> unitCardList) {
+    public UnitCardAdapter(
+            AppCompatActivity context,
+            Project project,
+            int chapter,
+            List<UnitCard> unitCardList,
+            ProjectDatabaseHelper db) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/adapters/UnitCardAdapter.java]: newline for ) {

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/DatabaseResyncTask.java [hunk 25]
```diff
@@ -23,13 +22,13 @@
  * Created by sarabiaj on 9/27/2016.
  */
 public class DatabaseResyncTask extends Task implements ProjectDatabaseHelper.OnLanguageNotFound, ProjectDatabaseHelper.OnCorruptFile {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/DatabaseResyncTask.java]: line length

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java [hunk 26]
```diff
@@ -80,7 +89,10 @@ public ProjectListResyncTask(int taskId, Context ctx, FragmentManager fm) {
                                                 for (int i = 0; i < c.length; i++) {
                                                     try {
                                                         WavFile wav = new WavFile(c[i]);
-                                                        mode = db.getMode(db.getModeId(wav.getMetadata().getModeSlug(), wav.getMetadata().getAnthology()));
+                                                        mode = db.getMode(db.getModeId(
+                                                                wav.getMetadata().getModeSlug(),
+                                                                wav.getMetadata().getAnthology()
+                                                        ));
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java]: code formatting

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java [hunk 27]
```diff
@@ -130,20 +148,20 @@ public ProjectListResyncTask(int taskId, Context ctx, FragmentManager fm) {
...
         Map<Project, File> directoriesOnFs = getProjectDirectoriesOnFileSystem();
         //if the number of projects doesn't match up between the filesystem and the db, OR,
         //the projects themselves don't match an id in the db, then resync everything (only resyncing
         // projects missing won't remove dangling take references in the db)
         //NOTE: removing a project only removes dangling takes, not the project itself from the db
-        if (directoriesOnFs.size() != db.getNumProjects() || db.projectsNeedingResync(directoriesOnFs.keySet()).size() > 0) {
-            fullResync(db, directoriesOnFs);
+        if (directoriesOnFs.size() != db.getNumProjects() || db.projectsNeedingResync(
+                directoriesOnFs.keySet()
+        ).size() > 0) {
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/ProjectManager/tasks/resync/ProjectListResyncTask.java]: I know you're meaning to break apart the method call in projects needing resync, but you want to make sure that it lines up vertically when you do that, especially with its beginning and the close parenthesis

```
        if (
            directoriesOnFs.size() != db.getNumProjects() 
                || db.projectsNeedingResync(directoriesOnFs.keySet()).size() > 0
         ) {
```
is probably what you would do here? but realistically, you'd rather assign variables

```
boolean projectCountDiffers = directoriesOnFs.size != db.getNumProjects;
boolean projectsNeedResync = db.projectsNeedingResync(directoriesOnFs.keySet()).size > 0
if(projectCountDiffers || projectsNeedResync) {
```

added benefit here is the if statement is more readable as to its intent

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectPlugin.java [hunk 28]
```diff
@@ -55,15 +55,19 @@ public ProjectPlugin(File pluginDir, File plugin) throws IOException {
...
         List<Version> versions = readVersions(new JsonReader(versionReader));
 //        Reader chunksReader = new FileReader(new File(pluginDir, "Chunks/" + chunksPath));
 //        readChunks(new JsonReader(chunksReader));
 
-        importPluginToDatabase(context, books.toArray(new Book[books.size()]), versions.toArray(new Version[versions.size()]), modes.toArray(new Mode[modes.size()]));
+        importPluginToDatabase(
+                books.toArray(new Book[books.size()]),
+                versions.toArray(new Version[versions.size()]),
+                modes.toArray(new Mode[modes.size()]),
+                db);
```
Comment (jsarabia) [inline translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/project/ProjectPlugin.java]: ); on new line

File: translationRecorder/app/src/main/AndroidManifest.xml (hunk 20 above)
Comment (mXaln) [reply translationRecorder/app/src/main/AndroidManifest.xml:99]: @sarabiaj Since Android api 24, in order for the app to give an access to it's content to other apps, this FileProvider is needed. You can try to share a project on devices with Android 7 and up, it will crash. 

https://developer.android.com/training/secure-file-sharing/setup-sharing
